File: framework/languages/node/packages/stream-connector/src/Contracts/IZlinkStreamConnector.ts
import type { Disposable } from './ZlinkStreamInterfaces';
import type { RequiredZlinkStreamConnectorOptions } from './ZlinkStreamConnectorOptions';
import type {
  ZlinkStreamRequestCall,
  ZlinkStreamSendCall,
  ZlinkStreamExpectNoneCall,
  ZlinkStreamSequenceCall,
  ZlinkStreamWaitCall
} from './Calls/ZlinkStreamCalls';
import type {
  ZlinkStreamConnectionStateChanged,
  ZlinkStreamEncodedPayload,
  ZlinkStreamError,
  ZlinkStreamMessage
} from './ZlinkStreamModels';
import type {
  ZlinkStreamCloseReason,
  ZlinkStreamConnectionState,
  ZlinkStreamDiagnosticsLevel
} from './ZlinkStreamEnums';
import type { ZlinkStreamActor } from './ZlinkStreamActor';

export interface ZlinkStreamConnector {
  readonly isConnected: boolean;
  readonly state: ZlinkStreamConnectionState;
  readonly closeReason?: ZlinkStreamCloseReason;
  readonly options: RequiredZlinkStreamConnectorOptions;
  readonly pendingDispatchCount: number;
  /**
   * Current diagnostics level (spec 26 §4.1, spec stream-connector 32 §13).
   * Always equal to `options.diagnosticsLevel`.
   */
  readonly diagnosticsLevel: ZlinkStreamDiagnosticsLevel;
  readonly actors: readonly ZlinkStreamActor[];
  actor(actorId: string): ZlinkStreamActor | undefined;
  onActorBound(
    handler: (actor: ZlinkStreamActor, signal?: AbortSignal) => Promise<void> | void
  ): Disposable;
  onActorUnbound(
    handler: (actor: ZlinkStreamActor, signal?: AbortSignal) => Promise<void> | void
  ): Disposable;
  /**
   * Changes the diagnostics level at runtime without recreating the
   * connector (spec 26 §4.1, spec stream-connector 32 §13). Applies to
   * processing points that read the level after this call returns; never
   * applied retroactively to frames already built. Rejects unknown values
   * with {@link import('./ZlinkStreamEnums').ZlinkStreamErrorCode.ConfigurationError}.
   */
  setDiagnosticsLevel(level: ZlinkStreamDiagnosticsLevel): void;
  /**
   * Asynchronous counterpart of {@link setDiagnosticsLevel} (spec
   * stream-connector 32 §13). Both surfaces change the same value, and this
   * one never replaces the synchronous surface: the synchronous call does not
   * wait for a completion, so a caller whose idiom is an awaited completion
   * uses this one instead. Rejects unknown values the same way.
   */
  setDiagnosticsLevelAsync(level: ZlinkStreamDiagnosticsLevel): Promise<void>;
  /**
   * Number of packets received under `name` on the current connection (spec
   * stream-connector 32 §10). Counts arrivals, so consuming a message through
   * a handler or a wait surface never lowers it, and the dispatch mode does
   * not change it. Restarts at 0 every time a connection is established.
   */
  receivedCount(name: string): number;
  onErrorReceived(
    handler: (error: ZlinkStreamError, signal?: AbortSignal) => Promise<void> | void
  ): Disposable;
  onDisconnected(handler: (signal?: AbortSignal) => Promise<void> | void): Disposable;
  onConnectionStateChanged(
    handler: (
      change: ZlinkStreamConnectionStateChanged,
      signal?: AbortSignal
    ) => Promise<void> | void
  ): Disposable;
  connect(signal?: AbortSignal): Promise<void>;
  close(signal?: AbortSignal): Promise<void>;
  dispatch(signal?: AbortSignal): Promise<void>;
  send(payload: unknown, messageType?: Function): ZlinkStreamSendCall;
  request(payload: unknown, messageType?: Function): ZlinkStreamRequestCall;
  on<TPayload = ZlinkStreamEncodedPayload>(
    name: string,
    handler: (message: ZlinkStreamMessage<TPayload>, signal?: AbortSignal) => Promise<void> | void,
    messageType?: Function
  ): Disposable;
  /**
   * Wait surfaces (spec stream-connector 32 §10.1). Each takes the packet name
   * the caller states explicitly, or the payload constructor the options'
   * `nameResolver` turns into that name — TypeScript types are erased at
   * runtime, so the type-driven path takes a constructor value.
   */
  waitFor<TPayload = ZlinkStreamEncodedPayload>(
    nameOrType: string | Function
  ): ZlinkStreamWaitCall<TPayload>;
  expectNone<TPayload = ZlinkStreamEncodedPayload>(
    nameOrType: string | Function
  ): ZlinkStreamExpectNoneCall<TPayload>;
  waitForSequence<TPayload = ZlinkStreamEncodedPayload>(
    nameOrType: string | Function
  ): ZlinkStreamSequenceCall<TPayload>;
}

File: framework/languages/dotnet/src/Systems.Zlink.Stream.Connector/Contracts/IZlinkStreamConnector.cs
namespace Systems.Zlink.Stream.Connector.Contracts;

/// <summary>
///     Represents a client-side stream connection to a ZLink stream endpoint.
/// </summary>
/// <remarks>
///     <see cref="On" /> is the normal callback path for long-lived push handling.
///     <see cref="WaitFor" /> is a
///     deterministic wait path for samples, command-line flows, and e2e scenario
///     tests.
/// </remarks>
public interface IZlinkStreamConnector : IAsyncDisposable
{
    /// <summary>
    ///     Gets whether the connector is currently connected.
    /// </summary>
    bool IsConnected { get; }

    /// <summary>
    ///     Gets the current connection state.
    /// </summary>
    ZlinkStreamConnectionState State { get; }

    /// <summary>
    ///     Gets the reason the connection last ended, or <see langword="null" /> when this
    ///     connector has never lost a connection.
    /// </summary>
    /// <remarks>
    ///     Readable at any time, so code that missed the disconnect event still reads the
    ///     same value (stream-connector spec §6.2, .NET spec §10). A first connect that
    ///     fails also leaves a reason here — <see cref="ZlinkStreamErrorCode.ConnectTimeout" />
    ///     and <see cref="ZlinkStreamErrorCode.TlsValidationFailed" /> record
    ///     <see cref="ZlinkStreamCloseReason.TransportError" />. Reconnecting does not clear
    ///     the value; it keeps the reason of the last close.
    /// </remarks>
    ZlinkStreamCloseReason? CloseReason { get; }

    /// <summary>Gets the options used by this connector.</summary>
    ZlinkStreamConnectorOptions Options { get; }

    /// <summary>
    ///     Gets the number of messages waiting for manual dispatch.
    /// </summary>
    int PendingDispatchCount { get; }

    IReadOnlyList<IZlinkStreamActor> Actors { get; }

    IZlinkStreamActor? Actor(string actorId);

    IDisposable OnActorBound(Func<IZlinkStreamActor, CancellationToken, ValueTask> handler);

    IDisposable OnActorUnbound(Func<IZlinkStreamActor, CancellationToken, ValueTask> handler);

    /// <summary>
    ///     Starts the stream connection lifecycle operation.
    /// </summary>
    IZlinkStreamLifecycleCall Connect { get; }

    /// <summary>
    ///     Starts the stream close lifecycle operation.
    /// </summary>
    IZlinkStreamLifecycleCall Close { get; }

    /// <summary>
    ///     Starts a manual dispatch lifecycle operation.
    /// </summary>
    IZlinkStreamLifecycleCall Dispatch { get; }

    /// <summary>
    ///     Registers a handler invoked when the endpoint sends an error message.
    /// </summary>
    /// <remarks>
    ///     Connection events are registration methods rather than C# events because an
    ///     <c>event</c> hands back nothing to unsubscribe with (stream-connector spec §7,
    ///     .NET spec §3). Dispose the returned registration to remove the handler;
    ///     disposing it twice is not an error, and a removed handler runs on no later
    ///     dispatch.
    /// </remarks>
    IDisposable OnErrorReceived(Func<ZlinkStreamError, CancellationToken, ValueTask> handler);

    /// <summary>
    ///     Registers a handler invoked after the connector becomes disconnected.
    /// </summary>
    /// <remarks>
    ///     Dispose the returned registration to remove the handler.
    /// </remarks>
    IDisposable OnDisconnected(Func<ZlinkStreamDisconnected, CancellationToken, ValueTask> handler);

    /// <summary>
    ///     Registers a handler invoked when the connection state changes.
    /// </summary>
    /// <remarks>
    ///     Dispose the returned registration to remove the handler.
    /// </remarks>
    IDisposable OnConnectionStateChanged(
        Func<ZlinkStreamConnectionStateChanged, CancellationToken, ValueTask> handler
    );

    IDisposable OnRequestSending(Action<ZlinkStreamRequestSendingContext> handler);

    IDisposable OnReplyReceived(
        Func<ZlinkStreamReplyReceivedContext, CancellationToken, ValueTask> handler
    );

    /// <summary>
    ///     Gets the number of messages received with <paramref name="name" /> on the
    ///     current connection.
    /// </summary>
    /// <remarks>
    ///     The value counts arrivals, not retained messages: consuming a message through
    ///     a wait surface or dispatching it to an <see cref="On" /> handler leaves the
    ///     count unchanged, and the count is the same in both dispatch modes because it
    ///     advances when the packet arrives (stream-connector spec §10). The baseline is
    ///     the moment a connection is established — every reconnect starts again at zero.
    ///     The value serves scenario assertions and diagnostics, never flow control.
    /// </remarks>
    int ReceivedCount(string name);

    /// <summary>
    ///     Starts a send operation for an encoded payload.
    /// </summary>
    IZlinkStreamSendCall Send(ZlinkStreamEncodedPayload payload);

    /// <summary>
    ///     Starts a request operation for an encoded payload.
    /// </summary>
    IZlinkStreamRequestCall Request(ZlinkStreamEncodedPayload payload);

    /// <summary>
    ///     Registers a callback for messages with the given packet name.
    /// </summary>
    /// <remarks>
    ///     The returned registration removes the callback when disposed. This is the
    ///     normal API for production push handling.
    /// </remarks>
    IDisposable On(
        string name,
        Func<ZlinkStreamMessage<ZlinkStreamEncodedPayload>, CancellationToken, ValueTask> handler
    );

    /// <summary>
    ///     Starts a wait operation for the next unread received message with the given packet name.
    /// </summary>
    /// <remarks>
    ///     The matched message is consumed by this wait operation. Use this API for
    ///     deterministic sample, CLI, or e2e scenario flow; production clients
    ///     should normally use <see cref="On" />.
    /// </remarks>
    IZlinkStreamWaitCall WaitFor(string name);

    /// <summary>
    ///     Verifies that no unread message with the given packet name arrives during a configured window.
    /// </summary>
    IZlinkStreamExpectNoneCall ExpectNone(string name);

    /// <summary>
    ///     Waits for messages with the given packet name and verifies their arrival order.
    /// </summary>
    IZlinkStreamSequenceCall WaitForSequence(string name);
}

public interface IZlinkStreamActor
{
    string ActorId { get; }

    bool IsBound { get; }

    IZlinkStreamSendCall Send(ZlinkStreamEncodedPayload payload);

    IZlinkStreamRequestCall Request(ZlinkStreamEncodedPayload payload);

    IDisposable On(
        string name,
        Func<ZlinkStreamMessage<ZlinkStreamEncodedPayload>, CancellationToken, ValueTask> handler
    );
}
